{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "T0Eq7RHxyeeh",
        "outputId": "063cf14e-e4f1-45eb-897f-76b48aa6bca0"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: pandas in /usr/local/lib/python3.12/dist-packages (2.2.2)\n",
            "Collecting biopython\n",
            "  Downloading biopython-1.85-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: requests in /usr/local/lib/python3.12/dist-packages (2.32.4)\n",
            "Collecting pypdb\n",
            "  Downloading pypdb-2.4-py3-none-any.whl.metadata (3.0 kB)\n",
            "Requirement already satisfied: numpy>=1.26.0 in /usr/local/lib/python3.12/dist-packages (from pandas) (2.0.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.12/dist-packages (from pandas) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.12/dist-packages (from pandas) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.12/dist-packages (from pandas) (2025.2)\n",
            "Requirement already satisfied: charset_normalizer<4,>=2 in /usr/local/lib/python3.12/dist-packages (from requests) (3.4.3)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.12/dist-packages (from requests) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.12/dist-packages (from requests) (2.5.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.12/dist-packages (from requests) (2025.8.3)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.12/dist-packages (from python-dateutil>=2.8.2->pandas) (1.17.0)\n",
            "Downloading biopython-1.85-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m23.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading pypdb-2.4-py3-none-any.whl (40 kB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m40.6/40.6 kB\u001b[0m \u001b[31m2.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython, pypdb\n",
            "Successfully installed biopython-1.85 pypdb-2.4\n"
          ]
        }
      ],
      "source": [
        "!pip install pandas biopython requests pypdb\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Kinase information"
      ],
      "metadata": {
        "id": "hdqU5cQq0u86"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "http://www.kinhub.org/kinases.html"
      ],
      "metadata": {
        "id": "3ZdLHrTH0oGH"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import SwissProt\n",
        "from Bio import PDB\n",
        "from urllib.request import urlopen\n",
        "from io import StringIO\n",
        "import pandas as pd\n",
        "import tqdm\n",
        "import time\n",
        "\n",
        "def get_pdb_ids_from_uniprot(uniprot_id):\n",
        "    url = f\"https://rest.uniprot.org/uniprotkb/{uniprot_id}.txt\"\n",
        "    response = urlopen(url)\n",
        "    record_string_list = response.read().decode(\"utf-8\").split('\\n')\n",
        "    time.sleep(0.1)\n",
        "\n",
        "    pdb_ids = []\n",
        "    for dbref in record_string_list:\n",
        "        if 'PDB;' in dbref:\n",
        "            pdb_id = dbref.split(\"PDB;\")[1].strip()\n",
        "            pdb_ids.append(pdb_id)\n",
        "    return pdb_ids\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    df = pd.read_csv('./DrugDiscovery/KinaseData.csv')\n",
        "    structure_info_list = []\n",
        "\n",
        "    for index, row in tqdm.tqdm(df.iterrows(), total=len(df)):\n",
        "        uniprot_id = row['UniprotID']\n",
        "        try:\n",
        "            pdb_ids = get_pdb_ids_from_uniprot(uniprot_id)\n",
        "        except:\n",
        "            pdb_ids = []\n",
        "\n",
        "        if len(pdb_ids) > 0:\n",
        "            structure_info_list.append(True)\n",
        "        else:\n",
        "            structure_info_list.append(False)\n",
        "\n",
        "    df['Structure Info'] = structure_info_list\n",
        "    df.to_csv('./KinaseData_StructureInfo.csv', index=False)\n",
        "    print(df)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 374
        },
        "id": "WOqNJZeHPSco",
        "outputId": "be7c0de7-7c21-43e6-bfa6-b2c4dc7314c5"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "error",
          "ename": "FileNotFoundError",
          "evalue": "[Errno 2] No such file or directory: './DrugDiscovery/KinaseData.csv'",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m/tmp/ipython-input-3050896158.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m__name__\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"__main__\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 23\u001b[0;31m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'./DrugDiscovery/KinaseData.csv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     24\u001b[0m     \u001b[0mstructure_info_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1024\u001b[0m     \u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwds_defaults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1025\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1026\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1027\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1028\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    618\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    619\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 620\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    621\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    622\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1618\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1619\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandles\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIOHandles\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1620\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1621\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1622\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1878\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0;34m\"b\"\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1879\u001b[0m                     \u001b[0mmode\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;34m\"b\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1880\u001b[0;31m             self.handles = get_handle(\n\u001b[0m\u001b[1;32m   1881\u001b[0m                 \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1882\u001b[0m                 \u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.12/dist-packages/pandas/io/common.py\u001b[0m in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    871\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencoding\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;34m\"b\"\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    872\u001b[0m             \u001b[0;31m# Encoding\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 873\u001b[0;31m             handle = open(\n\u001b[0m\u001b[1;32m    874\u001b[0m                 \u001b[0mhandle\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    875\u001b[0m                 \u001b[0mioargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: './DrugDiscovery/KinaseData.csv'"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### *In order not to overload the PubTator3 server, we ask that users post no more than three requests per second."
      ],
      "metadata": {
        "id": "267MOcR0XjEm"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import requests\n",
        "import json\n",
        "import time\n",
        "import tqdm\n",
        "\n",
        "def get_disease_info_from_pubtator(gene):\n",
        "    # URL\n",
        "    url = \"https://www.ncbi.nlm.nih.gov/research/pubtator3-api/relations?e1=@GENE_%s&type=stimulate&e2=Disease\"%(gene)\n",
        "    # API\n",
        "    response = requests.get(url)\n",
        "\n",
        "\n",
        "    data = response.json() # JSON 응답 파싱\n",
        "    # 'source'에서 질병 이름만 추출\n",
        "    diseases = []\n",
        "    for each_data in data:\n",
        "        each_disease = each_data['source']\n",
        "        each_disease = each_disease.replace('@DISEASE_', '')\n",
        "        each_disease = each_disease.replace('_', ' ')\n",
        "        diseases.append(each_disease)\n",
        "\n",
        "    diseases = list(set(diseases))\n",
        "\n",
        "    return diseases\n",
        "\n",
        "df = pd.read_csv('./KinaseData_StructureInfo.csv')\n",
        "df_wo_structure = df[df['Structure Info']==False]\n",
        "\n",
        "disease_list = []\n",
        "all_disease_list = []\n",
        "for index, each_df in tqdm.tqdm(df_wo_structure.iterrows(), total=len(df_wo_structure)):\n",
        "    gene = each_df['HGNC Name']\n",
        "\n",
        "    try:\n",
        "        diseases = get_disease_info_from_pubtator(gene)\n",
        "        all_disease_list+=diseases\n",
        "        disease_string = ';'.join(diseases)\n",
        "    except:\n",
        "        disease_string = ''\n",
        "    disease_list.append(disease_string)\n",
        "\n",
        "\n",
        "df_wo_structure['Disease Information'] = disease_list\n",
        "df_wo_structure.to_csv('./KinaseData_WO_StructureInfo_DiseaseInfo.csv')\n",
        "\n",
        "all_disease_list = list(set(all_disease_list))\n",
        "for disease in all_disease_list:\n",
        "    print(disease)\n"
      ],
      "metadata": {
        "id": "XqQdYK7cTzgf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Which diseases below are not genetic and are considered rare diseases, or not genetic and have a high unmet medical need?\n",
        "\n",
        "Inflammation\n",
        "Tertiary Lymphoid Structures\n",
        "Diabetes Mellitus\n",
        "Reperfusion Injury\n",
        "Transfusion Reaction\n",
        "Arthropathy progressive pseudorheumatoid of childhood\n",
        "Osteosarcoma\n",
        "Hereditary Breast and Ovarian Cancer Syndrome\n",
        "Erythema Infectiosum\n",
        "Death\n",
        "Edema\n",
        "Radiation Injuries\n",
        "Noble Bass Sherman syndrome\n",
        "Diabetic Retinopathy\n",
        "Intracranial Hemorrhages\n",
        "Enteropathy Associated T Cell Lymphoma\n",
        "Long QT Syndrome\n",
        "Oto palato digital syndrome type 2\n",
        "Anemia Hemolytic\n",
        "Pulmonary Fibrosis\n",
        "Basal Ganglia Diseases\n",
        "Silicosis\n",
        "Cystitis\n",
        "Hypothermia\n",
        "Lymphoma B Cell\n",
        "Carcinoma Renal Cell\n",
        "Fibrosis\n",
        "Prostatic Neoplasms Castration Resistant\n",
        "Alveolitis Extrinsic Allergic\n",
        "Neoplastic Syndromes Hereditary\n",
        "Anemia Aplastic\n",
        "Neurobehavioral Manifestations\n",
        "Urinary Bladder Neoplasms\n",
        "Subarachnoid Hemorrhage\n",
        "Hyperglycemia\n",
        "Idiopathic Noncirrhotic Portal Hypertension\n",
        "Breast Neoplasms\n",
        "Leber Congenital Amaurosis\n",
        "Osteoglophonic dwarfism\n",
        "Calcinosis\n",
        "Ring Chromosome 20 Syndrome\n",
        "Enterocolitis Necrotizing\n",
        "Sepsis\n",
        "Asthma\n",
        "Autism Spectrum Disorder\n",
        "Apical Hypertrophic Cardiomyopathy\n",
        "Wounds and Injuries\n",
        "Arthritis Psoriatic\n",
        "Anemia\n",
        "Cytokine Release Syndrome\n",
        "Lymphatic Metastasis\n",
        "Osteoarthritis\n",
        "Demyelinating Diseases\n",
        "Stress Disorders Post Traumatic\n",
        "Carcinoma Ductal\n",
        "Atherosclerosis\n",
        "Systemic Inflammatory Response Syndrome\n",
        "Uterine Cervical Neoplasms\n",
        "Arthritis\n",
        "Neoplasms\n",
        "Atrial Fibrillation\n",
        "Choline Deficiency\n",
        "Neurodegenerative Diseases\n",
        "Carcinoma Signet Ring Cell\n",
        "Pain\n",
        "Fever\n",
        "Esophageal Squamous Cell Carcinoma\n",
        "Microcystic adnexal carcinoma\n",
        "Pulmonary Embolism\n",
        "Hepatic Encephalopathy\n",
        "Skin Diseases\n",
        "Liver Failure\n",
        "Squamous Intraepithelial Lesions\n",
        "Ischemia\n",
        "Multiple Trauma\n",
        "Carcinoma Non Small Cell Lung\n",
        "Iron Overload\n",
        "Mitochondrial Myopathies\n",
        "Meningioma\n",
        "Insulin Resistance\n",
        "Depressive Disorder\n",
        "Rumination Syndrome\n",
        "Hypertension\n",
        "Prostatic Intraepithelial Neoplasia\n",
        "Sleep Deprivation\n",
        "Epilepsy Post Traumatic\n",
        "Nervous System Diseases\n",
        "Ischemic Attack Transient\n",
        "Renal Insufficiency\n",
        "Protoporphyria Erythropoietic\n",
        "Myocardial Ischemia\n",
        "Brain Injury Chronic\n",
        "Melanoma\n",
        "Squamous Cell Carcinoma of Head and Neck\n",
        "Gliosis\n",
        "Glomerulonephritis Membranous\n",
        "Leukemia Lymphocytic Chronic B Cell\n",
        "Hyperglycemic Hyperosmolar Nonketotic Coma\n",
        "Lymphoma Follicular\n",
        "Tarlov Cysts\n",
        "Chromosome Aberrations\n",
        "Wounds Stab\n",
        "Pulmonary Arterial Hypertension\n",
        "Cystic Fibrosis\n",
        "Epstein Barr Virus Infections\n",
        "Lung Neoplasms\n",
        "Keratosis Actinic\n",
        "Cerebral Infarction\n",
        "Mitochondrial cytopathy\n",
        "Burns\n",
        "Diabetic Nephropathies\n",
        "Miosis\n",
        "Neoplasm Metastasis\n",
        "Hyperplasia\n",
        "Nasopharyngeal Carcinoma\n",
        "Carcinoma Basal Cell\n",
        "Carcinoma Hepatocellular\n",
        "Multiple Myeloma\n",
        "Exercise Induced Allergies\n",
        "Jejunal Diseases\n",
        "Brain Ischemia\n",
        "Kidney Failure Chronic\n",
        "Neuralgia\n",
        "Acute Kidney Injury\n",
        "Recurrence\n",
        "Thymus Neoplasms\n",
        "Abortion Spontaneous\n",
        "Colitis\n",
        "Sciatic Neuropathy\n",
        "Anodontia\n",
        "Hepatitis Chronic\n",
        "Neurofibromatosis 1\n",
        "Parkinsonian Disorders\n",
        "Mastocytosis Systemic\n",
        "Adenocarcinoma of Lung\n",
        "Cell Transformation Neoplastic\n",
        "Carcinoma Pancreatic Ductal\n",
        "Status Asthmaticus\n",
        "Personality Disorders\n",
        "Pancreatic Neoplasms\n",
        "Glioma\n",
        "Cognition Disorders\n",
        "Sarcoma Kaposi\n",
        "Attention Deficit Disorder with Hyperactivity\n",
        "Diabetes Mellitus Type 1\n",
        "Carcinoma Squamous Cell\n",
        "Heart Diseases\n",
        "Diabetes Mellitus Type 2\n",
        "Carcinogenesis\n",
        "Harlequin type ichthyosis\n",
        "Leukemia Eosinophilic Acute\n",
        "Immunologic Deficiency Syndromes\n",
        "Infarction\n",
        "Small Cell Lung Carcinoma\n",
        "Adenomyosis\n",
        "Spherocytosis Type 1\n",
        "Mitochondrial Diseases\n",
        "Kidney Diseases\n",
        "Neurotoxicity Syndromes\n",
        "Brain Infarction\n",
        "Adenomatous Polyps\n",
        "Preeclamptic toxemia\n",
        "Myocardial Infarction\n",
        "Nerve Degeneration\n",
        "Mullerian aplasia\n",
        "Status Epilepticus\n",
        "Myocardial Reperfusion Injury\n",
        "Gallbladder Neoplasms\n",
        "Obesity\n",
        "Calcinosis Cutis\n",
        "Lupus Erythematosus Systemic\n",
        "Hepatitis B\n",
        "Carcinoma Ovarian Epithelial\n",
        "Hyperoxia\n",
        "Neoplasms Basal Cell\n",
        "Necrosis\n",
        "Adenoma Oxyphilic\n",
        "Vasculitis\n",
        "Lung Injury\n",
        "Carcinoma Embryonal\n",
        "Ureteral Obstruction\n",
        "Seizures\n",
        "Esophageal Neoplasms\n",
        "Ventricular Remodeling\n",
        "Chromosomal Instability\n",
        "Colitis Ulcerative\n",
        "Parkinson Disease Secondary\n",
        "Glucose Intolerance\n",
        "Fibromyalgia\n",
        "Glioblastoma\n",
        "Fetal Diseases\n",
        "Arthritis Rheumatoid\n",
        "Magnesium Deficiency\n",
        "Escherichia coli Infections\n",
        "Non alcoholic Fatty Liver Disease\n",
        "Brain Edema\n",
        "Cognitive Dysfunction\n",
        "Blindness\n",
        "Myelitis Transverse\n",
        "Cone Dystrophy\n",
        "Anemia Sickle Cell\n",
        "Hypoxia Ischemia Brain\n",
        "Myopathies Structural Congenital\n",
        "Leukemia Promyelocytic Acute\n",
        "Renal Insufficiency Chronic\n",
        "Factor X Deficiency\n",
        "Cleft Palate\n",
        "Fetal Growth Retardation\n",
        "Cardiotoxicity\n",
        "Pulmonary Disease Chronic Obstructive\n",
        "Lipoma\n",
        "Hepatitis C\n",
        "Hypoxia Brain\n",
        "Ventricular Dysfunction Left\n",
        "Liver Neoplasms\n",
        "Post Acute COVID 19 Syndrome\n",
        "Drug Related Side Effects and Adverse Reactions\n",
        "Scleroderma Systemic\n",
        "Virus Diseases\n",
        "Arthritis Experimental\n",
        "Vascular Diseases\n",
        "Mouth Diseases\n",
        "Triple Negative Breast Neoplasms\n",
        "Brain Concussion\n",
        "Corneal Endothelial Cell Loss\n",
        "Prodromal Symptoms\n",
        "Heart Failure\n",
        "Hyper IgM Immunodeficiency Syndrome\n",
        "Adamantinoma\n",
        "Adenoma\n",
        "Hypoglycemia\n",
        "Colorectal Neoplasms\n",
        "Retinal Degeneration\n",
        "Weight Gain\n",
        "Huntington Disease\n",
        "Psoriasis\n",
        "Reticulocytosis\n",
        "Bites and Stings\n",
        "Ovarian Neoplasms\n",
        "Cystitis Interstitial\n",
        "Hyperinsulinism\n",
        "Retinitis\n",
        "Prostatic Neoplasms\n",
        "Idiopathic Pulmonary Fibrosis\n",
        "Malaria Cerebral\n",
        "Arthralgia\n",
        "Mandibular Nerve Injuries\n",
        "Myotonic Dystrophy\n",
        "Benign non infected urachal cyst\n",
        "Esophagitis\n",
        "Stomach Neoplasms\n",
        "Epilepsy Partial with Variable Foci\n",
        "Rectal Neoplasms\n",
        "Lens Diseases\n",
        "Leukemia Myeloid\n",
        "Leukemia\n",
        "Autoimmune Diseases\n",
        "Brain Injuries Traumatic\n",
        "16p11.2 Deletion Syndrome\n",
        "Micrognathism\n",
        "Behcet Syndrome\n",
        "Retinal Detachment\n",
        "Pre Eclampsia\n",
        "Neutrophil Actin Dysfunction\n",
        "Leukopenia\n",
        "Celiac Disease\n",
        "Malformations of Cortical Development Group I\n",
        "Syringoma\n",
        "Hydrops Fetalis\n",
        "Hemorrhage\n",
        "613882\n",
        "Memory Disorders\n",
        "Uterine Cervicitis\n",
        "Leukemia Biphenotypic Acute\n",
        "Hypoxia\n",
        "Infections\n",
        "Psychological Distress\n",
        "Pancreatitis\n",
        "Mosaic variegated aneuploidy syndrome\n",
        "248310\n",
        "Arthritis Infectious\n",
        "Hyperhomocysteinemia\n",
        "Cardiomyopathy Hypertrophic\n",
        "Cardiomyopathy Dilated\n",
        "Mastitis\n",
        "Crohn Disease\n",
        "Prostatic Hyperplasia\n",
        "Seminoma\n",
        "Bone Diseases\n",
        "Infarction Middle Cerebral Artery\n",
        "Adenocarcinoma\n",
        "Abdominal Injuries\n",
        "Learning Disabilities\n",
        "Hyperimmunoglobulin G1(A1) Syndrome\n",
        "Parkinson Disease"
      ],
      "metadata": {
        "id": "mg_k0Y2EW30N"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import pandas as pd\n",
        "from scipy.stats import shapiro\n",
        "\n",
        "population = np.random.exponential(scale=1.0, size=10000)\n",
        "sample_means = [np.mean(np.random.choice(population, size=30)) for _ in range(1000)]\n",
        "\n",
        "pop_stat, pop_p = shapiro(population[:5000])\n",
        "sample_stat, sample_p = shapiro(sample_means)\n",
        "\n",
        "fig, axes = plt.subplots(1, 2, figsize=(14, 5))\n",
        "\n",
        "axes[0].hist(population, bins=50, density=True, alpha=0.7, color='skyblue')\n",
        "axes[0].set_title(f'Population (Exponential)')\n",
        "axes[0].set_xlabel('Value')\n",
        "axes[0].set_ylabel('Density')\n",
        "\n",
        "axes[1].hist(sample_means, bins=50, density=True, alpha=0.7, color='salmon')\n",
        "axes[1].set_title(f'Sampling Distribution of the Mean')\n",
        "axes[1].set_xlabel('Sample Mean')\n",
        "axes[1].set_ylabel('Density')\n",
        "\n",
        "plt.tight_layout()\n",
        "plt.show()\n",
        "print(pop_p)\n",
        "print(sample_p)"
      ],
      "metadata": {
        "id": "XabvMEY2mMkW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# 1000번의 반복 실험을 통해 t-value를 계산하여 분포 시각화\n",
        "t_values = []\n",
        "\n",
        "for _ in range(1000):\n",
        "    group1 = np.random.normal(loc=0, scale=1, size=30)\n",
        "    group2 = np.random.normal(loc=0.5, scale=1.2, size=30)\n",
        "    t_statistic, _ = ttest_ind(group1, group2)\n",
        "    t_values.append(t_statistic)\n",
        "\n",
        "# t-value 분포 시각화\n",
        "plt.figure(figsize=(10, 6))\n",
        "plt.hist(t_values, bins=30, density=True, alpha=0.7, color='lightgreen', edgecolor='black')\n",
        "\n",
        "# 세부설정\n",
        "plt.title('Distribution of t-values from 1000 Simulations')\n",
        "plt.xlabel('t-value')\n",
        "plt.ylabel('Density')\n",
        "plt.legend()\n",
        "plt.grid(True)\n",
        "plt.show()"
      ],
      "metadata": {
        "id": "4PCdprkMmLA-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!wget --no-check-certificate \"https://www.dropbox.com/scl/fi/533wpky7dz2kvcjqm4jjx/lec5_data.zip?rlkey=b2cko74izyin5wr05z3d99svw&st=x2rxrzha&dl=0\" -O \"lec5_data.zip\"\n",
        "!unzip lec5_data.zip"
      ],
      "metadata": {
        "id": "63_Xu2NF-AKE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "cellline_df = pd.read_csv(\"./Model.csv\")\n",
        "cellline_df"
      ],
      "metadata": {
        "id": "R6dO9Bm6-CO0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "ess_prob_df = pd.read_csv(\"./CRISPRGeneDependency.csv\", index_col=0)\n",
        "ess_prob_df"
      ],
      "metadata": {
        "id": "7a8ZC3q2-Epo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "mutation_df = pd.read_csv(\"./mutations_NAsdropped.csv\", index_col=0)\n",
        "mutation_df"
      ],
      "metadata": {
        "id": "pqstDd3j-F-0"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from scipy.stats import ttest_ind\n",
        "\n",
        "def find_significant_genes(essentiality_df, control_group, test_group, threshold=0.001):\n",
        "    genes = essentiality_df.columns\n",
        "    control_df = essentiality_df.loc[control_group]\n",
        "    test_df = essentiality_df.loc[test_group]\n",
        "\n",
        "    significant_genes = []\n",
        "\n",
        "    for gene in genes:\n",
        "        control_prob = control_df[gene]\n",
        "        test_prob = test_df[gene]\n",
        "\n",
        "        control_mean = control_prob.mean()\n",
        "        test_mean = test_prob.mean()\n",
        "\n",
        "        if control_mean < 0.5 and test_mean > 0.5:\n",
        "            _, p_value = ttest_ind(control_prob, test_prob)\n",
        "\n",
        "            if p_value < threshold:\n",
        "                significant_genes.append(gene)\n",
        "\n",
        "    return significant_genes\n",
        "\n",
        "\n",
        "lof_mutations_df = breast_mutation_df[breast_mutation_df['likely_lof'] == True]\n",
        "lof_genes = lof_mutations_df['gene'].unique()\n",
        "essentiality_cell_lines = ess_prob_df.index.tolist()\n",
        "\n",
        "for gene, gene_mutation_df in lof_mutations_df.groupby('gene'):\n",
        "    lof_cell_lines = set(gene_mutation_df['depmap_id']) & set(essentiality_cell_lines)\n",
        "    non_lof_cell_lines = set(breast_cancer_cellline_ids) - lof_cell_lines\n",
        "    non_lof_cell_lines &= set(essentiality_cell_lines)\n",
        "\n",
        "    if len(lof_cell_lines) >= 3:\n",
        "        candidate_genes = find_significant_genes(\n",
        "            ess_prob_df,\n",
        "            control_group=list(non_lof_cell_lines),\n",
        "            test_group=list(lof_cell_lines)\n",
        "        )\n",
        "\n",
        "        print(f'Potential target genes for LOF gene: {gene}')\n",
        "        print(candidate_genes)"
      ],
      "metadata": {
        "id": "bTp2FhFq-KL4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "\n",
        "# Load Genomics of Drug Sensitivity in Cancer dataset\n",
        "df = pd.read_csv('./DrugDiscovery/Data-GDSC-OV_ANOVA.csv')\n",
        "\n",
        "# Select only PARP inhibitors\n",
        "df = df[df['Drug target'] == 'PARP1, PARP2']\n",
        "\n",
        "# Filter by p-value threshold (0.05)\n",
        "df = df[df['feature_pval'] <= 0.05]\n",
        "\n",
        "# Exclude features starting with 'cna' (copy number alteration)\n",
        "df = df[~df['Feature Name'].str.startswith('cna')]\n",
        "\n",
        "# Print unique feature names\n",
        "print('Feature name')\n",
        "print(list(set(df['Feature Name'])))\n",
        "\n",
        "# Get unique PARP inhibitor drug names\n",
        "Target_drugs = list(set(df['Drug name']))\n",
        "print('PARP inhibitors')\n",
        "print(Target_drugs)"
      ],
      "metadata": {
        "id": "FIfle5uk-f8G"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Define the target gene for this analysis\n",
        "target_gene = 'BRCA1'\n",
        "\n",
        "# Load genetic variant data for ovarian cancer cell lines from GDSC dataset\n",
        "df2 = pd.read_csv('./DrugDiscovery/Data-GDSC-OV_Genetic_features_variant.csv')\n",
        "\n",
        "# Print the loaded dataframe\n",
        "print(df2)\n",
        "\n",
        "# Filter for genetic feature information related to the target gene\n",
        "tmp_df = df2[df2['Genetic Feature'] == '%s_mut'%(target_gene)]\n",
        "\n",
        "# Extract cell lines with mutations in the target gene\n",
        "tmp_df_pos = tmp_df[tmp_df['IS Mutated'] == 1]\n",
        "\n",
        "# Extract cell lines without mutations in the target gene\n",
        "tmp_df_neg = tmp_df[tmp_df['IS Mutated'] == 0]\n",
        "\n",
        "# Get unique list of cell lines with mutations\n",
        "mut_pos_cells = list(set(tmp_df_pos['Cell Line Name']))\n",
        "\n",
        "# Get unique list of cell lines without mutations\n",
        "mut_neg_cells = list(set(tmp_df_neg['Cell Line Name']))\n",
        "\n",
        "# Print cell lines with and without mutations\n",
        "print(mut_pos_cells)\n",
        "print(mut_neg_cells)"
      ],
      "metadata": {
        "id": "MBlmKj0B-h1l"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "\n",
        "# Load IC50 data from CSV file\n",
        "ic50_df = pd.read_csv('./DrugDiscovery/Data-GDSC-OV_IC_Mon Sep 18 04_05_59 2023.csv')\n",
        "# Filter for selected inhibitors\n",
        "ic50_df = ic50_df[ic50_df['Drug Name'].isin(Target_drugs)]\n",
        "\n",
        "# Convert IC50 from log scale to linear scale by applying exponential\n",
        "ic50_df['IC50 new'] = np.exp(ic50_df['IC50'])\n",
        "\n",
        "# Analyze data for each PARP inhibitor drug by grouping\n",
        "for each_drug, each_df in ic50_df.groupby('Drug Name'):\n",
        "    # Extract data for cell lines with mutations\n",
        "    pos_data_df = each_df[each_df['Cell Line Name'].isin(mut_pos_cells)]\n",
        "\n",
        "    # Extract data for cell lines without mutations\n",
        "    neg_data_df = each_df[each_df['Cell Line Name'].isin(mut_neg_cells)]\n",
        "\n",
        "    # Get IC50 values for cell lines with mutations\n",
        "    pos_ic50_values = pos_data_df['IC50 new'].values\n",
        "\n",
        "    # Get IC50 values for cell lines without mutations\n",
        "    neg_ic50_values = neg_data_df['IC50 new'].values\n",
        "\n",
        "    # Create boxplot comparing IC50 values for mutation-positive and mutation-negative groups\n",
        "    plt.boxplot([pos_ic50_values, neg_ic50_values], labels=['%s mut positive'%(target_gene), '%s mut negative'%(target_gene)])\n",
        "    plt.title('%s'%(each_drug))\n",
        "    plt.xlabel('Group')\n",
        "    plt.ylabel('IC50 (uM)')\n",
        "    plt.show()"
      ],
      "metadata": {
        "id": "2diIfy5N-j8A"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import os\n",
        "import json\n",
        "import pandas as pd\n",
        "from Bio import Entrez\n",
        "from openai import OpenAI\n",
        "from typing import List, Dict\n",
        "import time\n",
        "import re\n",
        "\n",
        "# Set your contact email for NCBI Entrez and initialize OpenAI client\n",
        "Entrez.email = \"your_email@example.com\"\n",
        "openai_client = OpenAI(api_key=os.getenv(\"OPENAI_API_KEY\", \"input\")) #\n",
        "\n",
        "def fetch_pubmed_abstracts(disease: str, gene: str, max_results: int = 5) -> List[Dict]:\n",
        "    \"\"\"\n",
        "    Search PubMed for articles containing both disease and gene terms.\n",
        "    Returns a list of dicts with PMID, title, and abstract.\n",
        "    \"\"\"\n",
        "    try:\n",
        "        query = f\"{disease}[Title/Abstract] AND {gene}[Title/Abstract]\"\n",
        "        handle = Entrez.esearch(db=\"pubmed\", term=query, retmax=max_results)\n",
        "        record = Entrez.read(handle)\n",
        "        handle.close()\n",
        "        pmids = record[\"IdList\"]\n",
        "        results = []\n",
        "        for pmid in pmids:\n",
        "            handle = Entrez.efetch(db=\"pubmed\", id=pmid, retmode=\"xml\")\n",
        "            article = Entrez.read(handle)['PubmedArticle'][0]\n",
        "            handle.close()\n",
        "            title = article['MedlineCitation']['Article']['ArticleTitle']\n",
        "            abstract_list = article['MedlineCitation']['Article'].get('Abstract', {}).get('AbstractText', [])\n",
        "            abstract = abstract_list[0] if abstract_list else \"\"\n",
        "            results.append({\"pmid\": pmid, \"title\": title, \"abstract\": abstract})\n",
        "        return results\n",
        "    except Exception as e:\n",
        "        print(f\"Error fetching PubMed abstracts for {disease} and {gene}: {e}\")\n",
        "        return []\n",
        "\n",
        "def parse_openai_response(response_content):\n",
        "    # Use regex to find the first valid JSON object in the response\n",
        "    json_match = re.search(r'\\{.*\\}', response_content, re.DOTALL)\n",
        "    if json_match:\n",
        "        json_str = json_match.group(0)\n",
        "        try:\n",
        "            return json.loads(json_str)\n",
        "        except json.JSONDecodeError as e:\n",
        "            raise json.JSONDecodeError(f\"Failed to parse JSON: {e}\", json_str, e.pos)\n",
        "    else:\n",
        "        raise ValueError(\"No valid JSON found in response\")\n",
        "\n",
        "def evaluate_article(disease: str, gene: str, title: str, abstract: str) -> Dict:\n",
        "    prompt = f\"\"\"\n",
        "Disease: {disease}\n",
        "Gene: {gene}\n",
        "Article Title: {title}\n",
        "Abstract: {abstract}\n",
        "\n",
        "Instructions: Analyze the provided abstract carefully and respond in the specified JSON format. Focus *only* on information explicitly stated or strongly implied within this specific abstract.\n",
        "\n",
        "1.  **Summarize:** Provide a concise, one-sentence summary describing the *specific main finding* reported in the abstract concerning the relationship between the gene '{gene}' and the disease '{disease}'. **This summary must be derived directly from the abstract's content and should not be generic.**\n",
        "2.  **Identify Potential Therapeutic Target:** Determine if the abstract presents *specific findings* suggesting the gene '{gene}' is a potential therapeutic target for '{disease}'.\n",
        "    * Set `is_potential_target` to `true` **only if** the abstract describes results where:\n",
        "        * Modulating the gene's activity demonstrably affects disease pathology, or disease model outcomes.\n",
        "        **Crucially, this often involves evidence presented at the cellular (in vitro) or animal (in vivo) level, such as showing that *gene knock-down or knock-out leads to a demonstrable therapeutic effect* (e.g., reduced disease phenotype, improved cell viability, slowed disease progression in models).\n",
        "        ** Other forms of modulation (like pharmacological inhibition) showing similar effects also qualify.\n",
        "    * **Important:** Simply reporting a correlation (e.g., gene expression levels differ in patients) or describing the gene's general biological function is **not sufficient** to mark it as a potential target *unless* the abstract directly links this to a potential therapeutic intervention strategy supported by experimental evidence presented within the abstract. There must be a clear indication of therapeutic relevance derived from specific results.\n",
        "    * Base this determination strictly on the content of the provided abstract.\n",
        "3.  **Describe Mechanism (if applicable):**\n",
        "    * If `is_potential_target` is `true`, describe the *specific mechanism* mentioned or strongly implied *in the abstract* that explains *how* modulating '{gene}' could lead to a therapeutic benefit for '{disease}'.\n",
        "    * **This description must be based on details found within the abstract.** If the abstract supports `is_potential_target` being true but *does not* specify the underlying mechanism, state \"Mechanism not detailed in the abstract.\"\n",
        "    * If `is_potential_target` is `false`, leave the `mechanism` field as an empty string (\"\").\n",
        "    * **Do not use generic placeholders like '{{gene}}' or '{{disease}}' in your mechanism description.**\n",
        "\n",
        "Respond in strict JSON format with the following keys:\n",
        "- summary: [string] (The specific one-sentence summary based on the abstract)\n",
        "- is_potential_target: [boolean] (True only if the abstract provides specific findings supporting therapeutic potential, particularly evidence like knock-down/out effects in models, False otherwise)\n",
        "- mechanism: [string] (Specific mechanism from the abstract if `is_potential_target` is true and mechanism is described, \"Mechanism not detailed in the abstract.\" if true but not described, or \"\" if false)\n",
        "\n",
        "Example Output (Target with Mechanism - Knockdown):\n",
        "{{\n",
        "  \"summary\": \"The abstract demonstrates that siRNA-mediated knock-down of Gene Alpha in Disease X cell lines reduces cell proliferation and induces apoptosis.\",\n",
        "  \"is_potential_target\": true,\n",
        "  \"mechanism\": \"Knock-down of Gene Alpha disrupts the Survival Pathway Y, leading to reduced proliferation and apoptosis of Disease X cells, according to the abstract.\"\n",
        "}}\n",
        "\n",
        "Example Output (Target with Mechanism - Knockout):\n",
        "{{\n",
        "  \"summary\": \"This study shows that genetic knock-out of Gene Beta ameliorates disease symptoms and reduces pathological markers in a mouse model of Disease Z.\",\n",
        "  \"is_potential_target\": true,\n",
        "  \"mechanism\": \"The abstract suggests Gene Beta knock-out restores normal signaling pathway Q, which is dysregulated in Disease Z.\"\n",
        "}}\n",
        "\n",
        "Example Output (Target, Mechanism Not Detailed):\n",
        "{{\n",
        "  \"summary\": \"This abstract reports that pharmacological inhibition of Enzyme C significantly improved outcomes in an animal model of Disease D.\",\n",
        "  \"is_potential_target\": true,\n",
        "  \"mechanism\": \"Mechanism not detailed in the abstract.\"\n",
        "}}\n",
        "\n",
        "Example Output (Not a Target - Correlation Only):\n",
        "{{\n",
        "  \"summary\": \"The abstract describes a genome-wide association study identifying a SNP near Gene E that is associated with increased risk for Disease F.\",\n",
        "  \"is_potential_target\": false,\n",
        "  \"mechanism\": \"\"\n",
        "}}\n",
        "\n",
        "Example Output (Not a Target - Functional Study Only):\n",
        "{{\n",
        "  \"summary\": \"This research investigates the role of Protein G in regulating cellular trafficking pathways within neurons, without specific experimental links to Disease H treatment effects.\",\n",
        "  \"is_potential_target\": false,\n",
        "  \"mechanism\": \"\"\n",
        "}}\n",
        "    \"\"\"\n",
        "\n",
        "    response = openai_client.chat.completions.create(\n",
        "    model=\"gpt-4.1-mini\",  # Updated to a valid model; replace with desired model\n",
        "    messages=[\n",
        "        {\"role\": \"system\", \"content\": \"You are an expert biomedical researcher.\"},\n",
        "        {\"role\": \"user\", \"content\": prompt}\n",
        "    ],\n",
        "    temperature=0\n",
        "    )\n",
        "\n",
        "    content = response.choices[0].message.content.strip()\n",
        "\n",
        "    # Parse the cleaned JSON content\n",
        "    parsed_content = parse_openai_response(content)\n",
        "    return parsed_content\n",
        "\n",
        "def analyze_genes(disease: str, genes: List[str], max_articles_per_gene: int = 3) -> pd.DataFrame:\n",
        "    \"\"\"\n",
        "    Main function to fetch abstracts, evaluate each, and compile results.\n",
        "    Writes results to 'results.csv' and returns a pandas DataFrame.\n",
        "    \"\"\"\n",
        "    records = []\n",
        "    for gene in genes:\n",
        "        articles = fetch_pubmed_abstracts(disease, gene, max_results=max_articles_per_gene)\n",
        "        print('No. of articles : ', len(articles))\n",
        "        for art in articles:\n",
        "            #try:\n",
        "            ev = evaluate_article(disease, gene, art['title'], art['abstract'])\n",
        "            #except:\n",
        "            #    continue\n",
        "            records.append({\n",
        "                \"Disease\": disease,\n",
        "                \"Gene\": gene,\n",
        "                \"PMID\": art['pmid'],\n",
        "                \"Title\": art['title'],\n",
        "                \"IsPotentialTarget\": ev[\"is_potential_target\"],\n",
        "                \"Summary\": ev[\"summary\"],\n",
        "                \"Mechanism\": ev[\"mechanism\"]\n",
        "            })\n",
        "            time.sleep(0.5)\n",
        "\n",
        "    df = pd.DataFrame(records)\n",
        "    try:\n",
        "        df.to_csv(\"results.csv\", index=False)\n",
        "    except Exception as e:\n",
        "        print(f\"Error saving results to CSV: {e}\")\n",
        "    return df\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    disease_name = \"Alzheimer disease\"\n",
        "    candidate_genes = [\"APP\", \"PSEN1\", \"PSEN2\"]\n",
        "    df_results = analyze_genes(disease_name, candidate_genes, 3)\n",
        "    print(df_results)"
      ],
      "metadata": {
        "id": "OX347ztv-oIJ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import datamol as dm\n",
        "from molfeat.trans.base import MoleculeTransformer\n",
        "from molfeat.calc.pharmacophore import Pharmacophore2D\n",
        "from molfeat.trans.fp import FPVecTransformer\n",
        "\n",
        "smiles = \"O=C(NNc1cc(Br)c([O-])c(Br)c1)c1ccccc1\t\"\n",
        "# sanitize and standardize your molecules if needed\n",
        "\n",
        "transformer = FPVecTransformer(kind='ecfp', dtype=float)\n",
        "features = transformer(smiles)\n",
        "print('EFCP')\n",
        "print(features.shape)\n",
        "print(features)\n",
        "\n",
        "transformer = FPVecTransformer(\"maccs\", dtype=float)\n",
        "features = transformer(smiles)\n",
        "print('MACCS')\n",
        "print(features.shape)\n",
        "print(features)"
      ],
      "metadata": {
        "id": "l_4C_JxK_NDt"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}